### PR TITLE
feat(link): add support for negative and neutral states and better contrast on dark backgrounds FE-5339

### LINKED ISSUE #5407: Change link component colour
### Desired behaviour

Ability to pass the desired colour of the link to the component so it can be readable against different background colours.

### Current behaviour

Currently the Link component is only available in one green colour this is not always distinguishable against darker backgrounds therefore not meeting accessibility standards.

### Suggested Solution

_No response_

### CodeSandbox or Storybook URL

_No response_

### Anything else we should know?

_No response_

### Confidentiality

- [X] I confirm there is no confidential or commercially sensitive information included.

### PR DESCRIPTION
**Pull request by edleeks87**
fix #5407

### Proposed behaviour

Adds `variant` prop to support `negative` and `neutral` states. Adds `isDarkBackground` prop 
to allow for better colour contrast when `Link` is rendered on dark backgrounds

image

### Current behaviour

No support for `variant` prop and or improving contrast on dark backgrounds

### Checklist

- [x] Commits follow our style guide
- [x] Related issues linked in commit messages if required
- [x] Screenshots are included in the PR if useful
- [x] Unit tests added or updated if required
- [x] Storybook added or updated if required
- [x] Typescript `d.ts` file added or updated if required

#### QA

- [ ] Tested in CodeSandbox/storybook
- [ ] Add new Cypress test coverage if required
- [ ] Carbon implementation matches Design System/designs
- [ ] UI Tests GitHub check reviewed if required

### Additional context

DS are aware of contrast issues with `disabled` links and will review at a later date

### Testing instructions

The following CodeSandbox is an example of the broken behaviour.
You can see the new behaviour by looking at the version in the comment by `codesandbox[bot]`.

https://codesandbox.io/s/cold-voice-ou4642?file=/src/index.js

### CODE REVIEW

File: src/components/link/link.style.ts [hunk 1]
```diff
@@ -31,14 +55,32 @@ const StyledLink = styled.span<StyledLinkProps>`
...
+          : "var(--colorsActionMajor500)"};
+
+        &:hover {
+          color: ${variant === "dark"
+            ? "var(--colorsActionMajor450)"
+            : "var(--colorsActionMajor600)"};
+        }
+
+        &:focus {
+          background-color: var(--colorsSemanticFocus250);
```
Comment (Parsium) [inline src/components/link/link.style.ts:74]:
```suggestion
          background-color: var(--colorsSemanticFocus250);
          color: var(--colorsActionsMajorYin090);
```

File: src/components/link/link.stories.mdx [hunk 2]
```diff
@@ -184,6 +184,37 @@ perform client-side routing, but still provide the `href` to render an HTML link
   </Story>
 </Canvas>
 
+### On a dark background
```
Comment (Parsium) [inline src/components/link/link.stories.mdx:187]: Should we also include the disabled link as well in this story, so it matches the figma design we received?

File: src/components/link/link.style.ts [hunk 3]
```diff
@@ -56,26 +98,24 @@ const StyledLink = styled.span<StyledLinkProps>`
 
       &:hover {
         cursor: pointer;
-        color: ${isSkipLink
-          ? "var(--colorsUtilityYin090)"
-          : "var(--colorsActionMajor600)"};
 
         ${StyledIcon} {
-          color: var(--colorsActionMajor600);
+          color: ${variant === "dark"
```
Comment (Parsium) [inline src/components/link/link.style.ts:103]: I noticed in the design when the link with icon is focused, the icon colour should also change to `colorsActionMajorYin090` same as the text for both light and dark variants. Could you also include this please?

From DS:
Screenshot 2022-09-08 at 11 48 18

Currently:
Screenshot 2022-09-08 at 11 48 42

Screenshot 2022-09-08 at 11 49 13

File: src/components/link/link.spec.tsx [hunk 4]
```diff
@@ -293,4 +293,81 @@ describe("Link", () => {
...
+        {
+          color: "var(--colorsActionMajor450)",
+        },
+        wrapper,
+        { modifier: "a:hover" }
+      );
+
+      assertStyleMatch(
+        {
+          backgroundColor: "var(--colorsSemanticFocus250)",
```
Comment (Parsium) [inline src/components/link/link.spec.tsx:339]:
```suggestion
          backgroundColor: "var(--colorsSemanticFocus250)",
          color: "var(--colorsActionMajorYin090)",
```

File: src/components/link/link.style.ts (hunk 3 above)
Comment (edleeks87) [reply src/components/link/link.style.ts:103]: Bug exists in master but I'll fix in this PR

**Review by DipperTheDan — APPROVED** [review]

File: src/components/link/link.style.ts [hunk 5]
```diff
@@ -1,95 +1,210 @@
...
+      color = "var(--colorsSemanticNegative350)";
+      hoverColor = "var(--colorsSemanticNegative450)";
+    } else if (variant === "neutral") {
+      color = "var(--colorsActionMinor100)";
+    }
+
+    return {
+      color,
+      hoverColor,
+      disabledColor: "var(--colorsActionMajorYang030)",
```
Comment (Parsium) [inline src/components/link/link.style.ts:67]: The style changes to `Link` look good to me! Unfortunately we do have a colour contrast violation for the disabled link on a dark background @edleeks87. @harpalsingh could we use a different temporary value for now until the token is updated, or is there an alternative token we could use?

Screenshot 2022-09-12 at 13 45 38

**Review by Parsium — DISMISSED** [review]

File: src/components/link/link.style.ts (hunk 5 above)
Comment (edleeks87) [reply src/components/link/link.style.ts:67]: Good spot, sorry I meant to add something to the PR description. DS are aware of the issue and will be reviewing the approach at some point but for now have asked to implement as is defined in the figma file

**Review by Parsium — DISMISSED** [review]

**Review by DipperTheDan — DISMISSED** [review]

**Review by Parsium — DISMISSED** [review]

**Review by sianford — APPROVED** [review]
Tested in Safari, Chrome, Firefox and Edge. Firefox presentation of the `onDarkBackground` example with smaller vertical spacing is not an issue with the component. Testing passed ✅

**Review by DipperTheDan — APPROVED** [review]

**Pull request merged**